File: src/simulation/SimulationTask.js

const CELLS_PER_TASK = 16*1024;// Default: 16*1024

const TASK_CELL_TYPE_ORDER_ARR = [
	CELL_PROPERTIES.CONSTANT,
	CELL_PROPERTIES.COPY,
	// bitwise operators
	CELL_PROPERTIES.OR,
	CELL_PROPERTIES.XOR,
	CELL_PROPERTIES.NOT,
	CELL_PROPERTIES.AND,
	CELL_PROPERTIES.LSHIFT,
	CELL_PROPERTIES.RSHIFT,
	// arithmetic & comparison
	CELL_PROPERTIES.ADD,
	CELL_PROPERTIES.SUB,
	CELL_PROPERTIES.MULT,
	CELL_PROPERTIES.DIV,
	CELL_PROPERTIES.GTH,
	CELL_PROPERTIES.LTH,
	CELL_PROPERTIES.EQUALS,
	CELL_PROPERTIES.GEQ,
	CELL_PROPERTIES.LEQ,
];
const TASK_CELL_TYPE_ORDER_MAP = new Map(TASK_CELL_TYPE_ORDER_ARR.map((props, i) => [props.type, i]));


/*
	Simulation data is split into batches, i.e. "Tasks",
	where threads can update a small group of cells independently,
	and then produce some output data that the parent thread gathers
	when all tasks are complete. 
*/
class SimulationTask {
	// ==================================================
	// Structors
	// --------------------------------------------------
	
	constructor(ibeg, iend, link_count) {
		/* Cell index range */
		this.ibeg = ibeg;
		this.iend = iend;
		const cell_count = iend - ibeg;
		/* Cells that belong to this task. */
		this.cell_buffer = new SimulationCellBuffer(cell_count);
		/* Variable-length link lists that belong to cells, gathered into a single array. */
		this.link_buffer = new SimulationLinkBuffer(link_count);
		/* List of updates to send to cells outside of this task. */
		this.out_buffer = new SimulationUpdateBuffer(link_count);
		/* Index-associated Set of booleans for ensuring update-indices are only added once. */
		this.update_added = new Uint8Array(cell_count).fill(0);
		/* List of (local) cell indices to update. */
		this.update_stack = new Uint32Array(cell_count);
		/* Number of cells to update during next cycle. */
		this.update_count = 0;
		/* Performance metrics. */
		this.perf_n_updates = 0;
		this.perf_n_outputs = 0;
	}
	
	// ==================================================
	// Helpers.
	// --------------------------------------------------
	
	/* Update cell value and add to list of cells to update during next cycle. */
	pushCellUpdate(ind, tgt, val) {
		const x = ind - this.ibeg;// convert to local index.
		this.cell_buffer.setCellValue(x, tgt, val);
		if(this.update_added[x] !== 1) {
			this.update_added[x] = 1;
			this.update_stack[this.update_count++] = x;
		}
	}
	
	/* Set the new output value of a cell, and notify output targets. */
	applyCellOutputChange(cellInd, val) {
		const len = this.cell_buffer.get_links_len(cellInd);
		const ofs = this.cell_buffer.get_links_ofs(cellInd);
		// write changed value to cell.
		this.cell_buffer.setCellValue(cellInd, 0x0/*Cell.LINK_TARGET.OUTPUT*/, val);
		// push update-values for output targets.
		for(let k=0;k<len;k++) {
			const ind = this.link_buffer.get_ind(ofs+k);
			const tgt = this.link_buffer.get_tgt(ofs+k);
			if((this.ibeg <= ind) & (ind < this.iend))
					this.pushCellUpdate(ind, tgt, val);
			else	this.out_buffer.push_fast(ind, tgt, val);
		}
	}
	
	/* Initialize and propagate cell value. */
	initializeCellValue(ind, val) {
		// push cell update containing initial value.
		this.pushCellUpdate(ind, Cell.LINK_TARGET.OUTPUT, val);
		// special handling for constants: manually propagate (since constants do not update normally).
		const x = ind - this.ibeg;
		if(this.cell_buffer.get_cell_order(x) === TASK_CELL_TYPE_ORDER_MAP.get(CELL_PROPERTIES.CONSTANT.type)) {
			this.applyCellOutputChange(x, val);
		}
	}
	
	/* Empty out list of update-indices. */
	clearCellUpdateList() {
		// NOTE: per-update is O(N) whereas array.fill is O(LENGTH >= N).
		for(let x=0;x<this.update_count;x++) this.update_added[this.update_stack[x]] = 0;
		this.update_count = 0;
	}
	
	// ==================================================
	// Update Cycle.
	// --------------------------------------------------
	
	/* Perform an update cycle. */
	update() {
		
		if(this.update_count <= 0) return;
		
		const cell_buffer = this.cell_buffer;
		const link_buffer = this.link_buffer;
		const out_buffer = this.out_buffer;
		
		// ============================================================
		// Sort inputs.
		// ------------------------------------------------------------
		// sort update indices into buckets based on cell type (bucket sort).
		const num = this.update_count;
		const counts  = new Uint32Array(TASK_CELL_TYPE_ORDER_ARR.length);
		const offsets = new Uint32Array(TASK_CELL_TYPE_ORDER_ARR.length);
		const sorted_inds = new Uint32Array(num);
		
		for(let x=0;x<num;x++) {
			const ind = this.update_stack[x];
			const bucket = cell_buffer.get_cell_order(ind);
			counts[bucket]++;
		}
		
		let sum=0;
		for(let x=0;x<TASK_CELL_TYPE_ORDER_ARR.length;x++) { offsets[x]=sum; sum+=counts[x]; }
		
		for(let x=0;x<num;x++) {
			const ind = this.update_stack[x];
			const bucket = cell_buffer.get_cell_order(ind);
			sorted_inds[offsets[bucket]++] = ind;
		}
		
		// ============================================================
		// Compute values.
		// ------------------------------------------------------------
		// gather input values.
		const inputA_values = new Uint32Array(num);
		const inputB_values = new Uint32Array(num);
		const old_outputs   = new Uint32Array(num);
		const new_outputs   = new Uint32Array(num);
		for(let x=0;x<num;x++) {
			const ci = sorted_inds[x];
			old_outputs[x]   = cell_buffer.get_value_out(ci);
			inputA_values[x] = cell_buffer.get_value_ina(ci);
			inputB_values[x] = cell_buffer.get_value_inb(ci);
		}
		
		// compute output values.
		let x=0;
		let X=0;
		X+=counts[ 0];
		X+=counts[ 1];	for(;x<X;x++) { new_outputs[x] = inputA_values[x]; }
		
		X+=counts[ 2];	for(;x<X;x++) { new_outputs[x] =   inputA_values[x] |  inputB_values[x]; }
		X+=counts[ 3];	for(;x<X;x++) { new_outputs[x] =   inputA_values[x] ^  inputB_values[x]; }
		X+=counts[ 4];	for(;x<X;x++) { new_outputs[x] = ~(inputA_values[x]); }
		X+=counts[ 5];	for(;x<X;x++) { new_outputs[x] =   inputA_values[x] &  inputB_values[x]; }
		X+=counts[ 6];	for(;x<X;x++) { new_outputs[x] =   inputA_values[x] << inputB_values[x]; }
		X+=counts[ 7];	for(;x<X;x++) { new_outputs[x] =   inputA_values[x] >> inputB_values[x]; }
		
		X+=counts[ 8];	for(;x<X;x++) { new_outputs[x] =  inputA_values[x] +  inputB_values[x]; }
		X+=counts[ 9];	for(;x<X;x++) { new_outputs[x] =  inputA_values[x] -  inputB_values[x]; }
		X+=counts[10];	for(;x<X;x++) { new_outputs[x] =  inputA_values[x] *  inputB_values[x]; }
		X+=counts[11];	for(;x<X;x++) { new_outputs[x] =  inputA_values[x] /  Math.max(inputB_values[x], 0x1); }
		X+=counts[12];	for(;x<X;x++) { new_outputs[x] = (inputA_values[x] >  inputB_values[x]) * 0xffffffff; }
		X+=counts[13];	for(;x<X;x++) { new_outputs[x] = (inputA_values[x] <  inputB_values[x]) * 0xffffffff; }
		X+=counts[14];	for(;x<X;x++) { new_outputs[x] = (inputA_values[x] == inputB_values[x]) * 0xffffffff; }
		X+=counts[15];	for(;x<X;x++) { new_outputs[x] = (inputA_values[x] >= inputB_values[x]) * 0xffffffff; }
		X+=counts[16];	for(;x<X;x++) { new_outputs[x] = (inputA_values[x] <= inputB_values[x]) * 0xffffffff; }
		
		// ============================================================
		// Collect outputs.
		// ------------------------------------------------------------
		// collect changed outputs.
		this.clearCellUpdateList();
		out_buffer.clear();
		for(let i=counts[0];i<num;i++) {
			if(new_outputs[i] !== old_outputs[i]) {
				this.applyCellOutputChange(sorted_inds[i], new_outputs[i]);
			}
		}
		
		// gather remaining performance metrics.
		this.perf_n_updates += num;
		this.perf_n_outputs += out_buffer.count;
	}
	
};

File: src/simulation/SimulationTask.cpp
#include "./SimulationCell.cpp"
#include "./SimulationLink.cpp"
#include "./SimulationUpdate.cpp"
#include "./SimulationArrayMap.cpp"
#include "../content/Link.cpp"
#include "../content/CellTypes.cpp"
#include <cassert>

/*
	Simulation data is split into batches, i.e. "Tasks",
	where threads can update a small group of cells independently,
	and then produce some output data that the parent thread gathers
	when all tasks are complete. 
*/
struct SimulationTask {
	/* Cells that belong to this task. */
	Vector<SimulationCell>		cell_buffer;
	/* Link lists that belong to cells, gathered into a single array. */
	Vector<SimulationLink>		link_buffer;
	/* List of updates to send to cells outside of this task. */
	Vector<SimulationUpdate>	out_buffer;
	/* List of internal indices to update during next update cycle. */
	SimulationArrayMap			update_map;
	/* First index in shared cell data. */
	u32		cell_ibeg;
	u32 	cell_iend;
	/* Performance metrics. */
	u32		perf_n_updates;
	u32		perf_n_outputs;
	
	SimulationTask(
		const Vector<SimulationCell>& cellbuf,
		const Vector<SimulationLink>& linkbuf,
		u32 cell_ibeg,
		u32 cell_iend
	) {
		// count total number of items.
		u32 cell_count = cell_iend - cell_ibeg;
		u32 link_count = 0;
		for(u32 i=cell_ibeg;i<cell_iend;i++) link_count += cellbuf[i].links_len;
		// initialize buffers.
		this->cell_ibeg = cell_ibeg;
		this->cell_iend = cell_iend;
		this->cell_buffer.reserve(cell_count);
		this->link_buffer.reserve(link_count);
		this-> out_buffer.reserve(link_count);
		this->update_map.clear();
		this->perf_n_updates = 0;
		this->perf_n_outputs = 0;
		// copy cells to task-local buffer.
		for(u32 i=cell_ibeg;i<cell_iend;i++) {
			this->cell_buffer.push_back(cellbuf[i]);
			assert(cell_buffer[i-cell_ibeg].task_order < NUM_CELL_TYPES);
			//printf("INIT ORDER: %u\n", this->cell_buffer[i-cell_ibeg].task_order);
		}
		// copy links to task-local buffer, and update cell list pointers.
		for(u32 i=cell_ibeg;i<cell_iend;i++)  {
			const u32 len = cellbuf[i].links_len;
			const u32 ofs = cellbuf[i].links_ofs;
			const u32 x = i - cell_ibeg;
			this->cell_buffer[x].links_ofs = this->link_buffer.size();
			for(int k=0;k<len;k++) {
				this->link_buffer.push_back(linkbuf[ofs + k]);
			}
		}
	}
	
	bool shouldUpdate() { return this->update_map.count > 0; }
	u32  toLocalIndex(u32 ind) const { return ind - this->cell_ibeg; }
	u32 toGlobalIndex(u32 ind) const { return ind + this->cell_ibeg; }
	
	/* Update cell value and add to list of cells to update during next cycle. */
	void pushCellUpdate(u32 ind, u32 tgt, u32 val) {
		const u32 x = toLocalIndex(ind);
		this->cell_buffer[x].setValue(tgt, val);
		this->update_map.add(x);
	}
	void pushCellUpdate(SimulationUpdate& upd) {
		const u32 x = toLocalIndex(upd.get_ind());
		this->cell_buffer[x].setValue(upd.get_tgt(), upd.val);
		this->update_map.add(x);
	}

	void propagateOutputChange(u32 x, u32 val) {
		const u32 len = this->cell_buffer[x].links_len;
		const u32 ofs = this->cell_buffer[x].links_ofs;
		for(u32 k=0;k<len;k++) {
			const u32 ind = this->link_buffer[ofs+k].get_ind();
			const u32 tgt = this->link_buffer[ofs+k].get_tgt();
			const bool isLocal = (this->cell_ibeg <= ind) & (ind < this->cell_iend);
			if(isLocal)	this->pushCellUpdate(ind, tgt, val);
			else		this->out_buffer.emplace_back(ind, tgt, val);
		}
	}
	
	/* Initialize and propagate cell value. */
	void initializeCellValue(u32 ind, u32 val) {
		const u32 x = toLocalIndex(ind);
		if(this->cell_buffer[x].task_order == CELL_TYPES.CONSTANT.taskOrder) {
			this->cell_buffer[x].setValue(LINK_TARGETS.OUTPUT, val);
			this->propagateOutputChange(x, val);
		} else {
			this->pushCellUpdate(ind, LINK_TARGETS.OUTPUT, val);
			this->propagateOutputChange(x, val);
		}
	}

	/* Modify cell value post-initialization. */
	void modifyCellValue(u32 ind, u32 val) {
		const u32 x = toLocalIndex(ind);
		this->cell_buffer[x].setValue(LINK_TARGETS.OUTPUT, val);
		this->propagateOutputChange(x, val);
	}

	
	/* Macro for setting loop interval for given cell type. */
	void loopConfig(int offsets[], int counts[], int& beg, int& end, const int& order) {
		beg = offsets[order];
		end = offsets[order] + counts[order];
	} 
	
	/* Perform an update cycle. */
	void update() {
		// ============================================================
		// Sort inputs.
		// ------------------------------------------------------------
		//printf("task sort\n");

		// bucket-sort local cell indices based on cell type.
		int offsets[NUM_CELL_TYPES];
		int counts [NUM_CELL_TYPES];
		for(int i=0;i<NUM_CELL_TYPES;i++) offsets[i] = 0;
		for(int i=0;i<NUM_CELL_TYPES;i++) counts [i] = 0;
		
		const int num = this->update_map.count;
		int sorted_inds[num];
		
		//for(int i=0;i<num;i++) printf("stack: %i\n", update_map.stack[i]);
		//for(int i=0;i<cell_buffer.size();i++) printf("order: %i\n", cell_buffer[i].task_order);

		for(int i=0;i<num;i++) {
			const u32 x = this->update_map.stack[i];
			const u32 bucket = cell_buffer[x].task_order;
			assert(bucket <NUM_CELL_TYPES);
			counts[bucket]++;
		}
		
		int sum=0;
		for(int x=0;x<NUM_CELL_TYPES;x++) { offsets[x]=sum; sum+=counts[x]; }
		
		for(int i=0;i<num;i++) {
			const u32 x = this->update_map.stack[i];
			const u32 bucket = cell_buffer[x].task_order;
			sorted_inds[offsets[bucket]++] = x;
		}
		
		// reset offsets for later use.
		for(int x=0;x<NUM_CELL_TYPES;x++) { offsets[x] -= counts[x]; }
		
		// ============================================================
		// Compute values.
		// ------------------------------------------------------------
		// gather input values.
		u32 out_prev[num];
		u32 out[num];
		u32 ina[num];
		u32 inb[num];
		for(int i=0;i<num;i++) {
			const SimulationCell& cell = cell_buffer[sorted_inds[i]];
			out_prev[i]	= cell.values[0];
			ina[i]		= cell.values[1];
			inb[i]		= cell.values[2];
		}
		
		// compute output values.
		{
		int b=0;
		int e=0;
		loopConfig(offsets, counts, b, e, CELL_TYPES.CONSTANT.taskOrder);	for(int x=b;x<e;x++) { out[x] = out_prev[x]; }
		loopConfig(offsets, counts, b, e, CELL_TYPES.COPY	.taskOrder);	for(int x=b;x<e;x++) { out[x] = ina[x]; }
		// bitwise operators
		loopConfig(offsets, counts, b, e, CELL_TYPES.OR		.taskOrder);	for(int x=b;x<e;x++) { out[x] =   ina[x] |  inb[x]; }
		loopConfig(offsets, counts, b, e, CELL_TYPES.XOR	.taskOrder);	for(int x=b;x<e;x++) { out[x] =   ina[x] ^  inb[x]; }
		loopConfig(offsets, counts, b, e, CELL_TYPES.NOT	.taskOrder);	for(int x=b;x<e;x++) { out[x] = ~(ina[x]); }
		loopConfig(offsets, counts, b, e, CELL_TYPES.AND	.taskOrder);	for(int x=b;x<e;x++) { out[x] =   ina[x] &  inb[x]; }
		loopConfig(offsets, counts, b, e, CELL_TYPES.LSHIFT	.taskOrder);	for(int x=b;x<e;x++) { out[x] =   ina[x] << inb[x]; }
		loopConfig(offsets, counts, b, e, CELL_TYPES.RSHIFT	.taskOrder);	for(int x=b;x<e;x++) { out[x] =   ina[x] >> inb[x]; }
		// arithmetic & comparison
		loopConfig(offsets, counts, b, e, CELL_TYPES.ADD	.taskOrder);	for(int x=b;x<e;x++) { out[x] =  ina[x] +  inb[x]; }
		loopConfig(offsets, counts, b, e, CELL_TYPES.SUB	.taskOrder);	for(int x=b;x<e;x++) { out[x] =  ina[x] -  inb[x]; }
		loopConfig(offsets, counts, b, e, CELL_TYPES.MULT	.taskOrder);	for(int x=b;x<e;x++) { out[x] =  ina[x] *  inb[x]; }
		loopConfig(offsets, counts, b, e, CELL_TYPES.DIV	.taskOrder);	for(int x=b;x<e;x++) { out[x] =  ina[x] /  std::max<u32>(inb[x], 0x1); }
		loopConfig(offsets, counts, b, e, CELL_TYPES.GTH	.taskOrder);	for(int x=b;x<e;x++) { out[x] = (ina[x] >  inb[x]) ? 0xffffffff : 0x0; }
		loopConfig(offsets, counts, b, e, CELL_TYPES.LTH	.taskOrder);	for(int x=b;x<e;x++) { out[x] = (ina[x] <  inb[x]) ? 0xffffffff : 0x0; }
		loopConfig(offsets, counts, b, e, CELL_TYPES.EQUALS	.taskOrder);	for(int x=b;x<e;x++) { out[x] = (ina[x] == inb[x]) ? 0xffffffff : 0x0; }
		loopConfig(offsets, counts, b, e, CELL_TYPES.GEQ	.taskOrder);	for(int x=b;x<e;x++) { out[x] = (ina[x] >= inb[x]) ? 0xffffffff : 0x0; }
		loopConfig(offsets, counts, b, e, CELL_TYPES.LEQ	.taskOrder);	for(int x=b;x<e;x++) { out[x] = (ina[x] <= inb[x]) ? 0xffffffff : 0x0; }
		}
		
		// ============================================================
		// Collect outputs.
		// ------------------------------------------------------------
		//printf("task collect\n");

		// collect changed outputs.
		this->update_map.clear();
		this->out_buffer.clear();
		for(int i=0;i<num;i++) if(out[i] != out_prev[i]) {
			const u32 x = sorted_inds[i];
			const u32 val = out[i];
			this->cell_buffer[x].setValue(LINK_TARGETS.OUTPUT, val);
			this->propagateOutputChange(x, val);
		}
		
		// gather remaining performance metrics.
		this->perf_n_updates += num;
		this->perf_n_outputs += out_buffer.size();
		//printf("<> task updates: %i \n", num);
		//printf("<> task outputs: %lu\n", out_buffer.size());
		//printf("<> task uinputs: %i \n", this->update_map.count);
	}
};



